packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -2354,17 +2366,96 @@ Hardhat Network's forking functionality only works with blocks from at least spu
     );
   }
 
+  private async _applyStateOverrideSet(stateOverrideSet: StateOverrideSet) {
+    // Multiple state override set can be configured for different addresses, hence the loop
+    for (const [addrToOverride, stateOverrideOptions] of Object.entries(
+      stateOverrideSet
+    )) {
+      const address = new Address(toBuffer(addrToOverride));
+
+      await this._overrideBalanceAndNonce(address, stateOverrideOptions);
+      await this._overrideCode(address, stateOverrideOptions);
+      await this._overrideStateAndStateDiff(address, stateOverrideOptions);
+    }
+  }
+
+  private async _overrideBalanceAndNonce(
+    address: Address,
+    stateOverrideOptions: StateOverrideOptions
+  ) {
+    const MAX_NONCE = 18446744073709551615n; // The max allowed hex value is 8 bytes: 0xFFFFFFFFFFFFFFFF
+    const { balance, nonce } = stateOverrideOptions;
+
+    if (nonce !== undefined && nonce > MAX_NONCE) {
+      throw new InvalidInputError(
+        "The 'nonce' property should occupy a maximum of 8 bytes. The max allowed value is: 0xFFFFFFFFFFFFFFFF."
+      );
+    }
+
+    await this._stateManager.modifyAccountFields(address, {
+      balance,
+      nonce,
+    });
+  }
+
+  private async _overrideCode(
+    address: Address,
+    stateOverrideOptions: StateOverrideOptions
+  ) {
+    if (stateOverrideOptions.code === undefined) return;
+
+    await this._stateManager.putContractCode(
+      address,
+      stateOverrideOptions.code
+    );
+  }
+
+  private async _overrideStateAndStateDiff(
+    address: Address,
+    stateOverrideOptions: StateOverrideOptions
+  ) {
+    const { state, stateDiff } = stateOverrideOptions;
+
+    if (state === undefined && stateDiff === undefined) return;
+
+    if (state !== undefined && stateDiff !== undefined) {
+      throw new InvalidInputError(
+        `The properties 'state' and 'stateDiff' cannot be used simultaneously when configuring the state override set passed to the eth_call method.`
+      );
+    }
+
+    if (stateOverrideOptions.state !== undefined) {
+      // Clear all storage except specified properties which are override
+      await this._stateManager.clearContractStorage(address);
+    }
+
+    const tmpStateProperties = state ?? stateDiff;
+
+    for (const [storageAddr, value] of Object.entries(tmpStateProperties!)) {
```

Comment:
I understand why this non-null assertion makes sense, but I've seen so many bugs caused by it that I normally go out of my way to avoid them.

In this case, you can do something like this, which is way more verbose but it should work:

```js
    let newState;
    if (state !== undefined && stateDiff === undefined) {
      newState = state;
    } else if (state === undefined && stateDiff !== undefined) {
      newState = stateDiff;
    } else if (state === undefined && stateDiff === undefined) {
      // nothing to do
      return;
    } else {
      throw new InvalidInputError(
        `The properties 'state' and 'stateDiff' cannot be used simultaneously when configuring the state override set passed to the eth_call method.`
      );
    }
```